design_documents/accessing_emitted_events.md

```diff
@@ -0,0 +1,191 @@
+If provided events will not be emitted it will panic with a detailed message.
+
+`events` are the subsequence of all events emitted by the function, but you can also require function
+to return exactly those (and not more) events with the `expect_exact_events` cheatcode. 
+
+```cario
+fn expect_exact_events(events: Array<snforge_std::Event>)
+```
+
+It will panic if:
+- not all defined events were emitted
+- some other events where emitted
+
+### Usage example
+
+```cario
+#[starknet::interface]
+trait IHelloEvent<TContractState> {
+    fn emit_store_name(self: @TContractState, name: felt252);
+}
+
+mod HelloEvent {
+    // ...
+    
+    #[event]
+    #[derive(Drop, starknet::Event)]
+    enum Event {
+        StoredName: StoredName, 
+    }
+
+    #[derive(Drop, starknet::Event)]
+    struct StoredName {
+        #[key]
+        user: ContractAddress,
+        name: felt252
+    }
+    
+    #[external(v0)]
+    impl IHelloEventImpl of super::IHelloEvent<ContractState> {
+        fn emit_store_name(self: @ContractState, name: felt252) {
+            // ...
+            self.emit(Event::StoredName(StoredName { user: get_caller_address(), name: name }));
+        }
+    }
+}
+
+use snforge_std::expect_events;
+use snforge_std::Event;
+
+#[test]
+fn check_emitted_event() {
+    // ...
+	expect_events(array![Event { name: 'StoredName', keys: array![123], data: array![456] }]);
+    let res = contract.emit_store_name(...);  // if the event is not emitted it will panic
+
+    let res = contract.emit_store_name(...);  // expect_events does not work here
+    
+    expect_exact_events(array![Event { name: 'StoredName', keys: array![123], data: array![456] }]);
+    let res = contract.emit_store_name(...);  // function has to emit exactly those events defined in the array otherwise it will panic
+    // ...
+}
+```
+
+## `spy_events` cheatcode
+
+Another idea is to give users the highest possible customization. There would be a handler for extracting events emitted
+after the line it was created. Users would have to assert events manually (which could be more complex than the previous
+solution but would add more flexibility).
+
+Introduce `spy_events` cheatcode with the signature:
+
+```cario
+fn spy_events() -> snforge_std::EventSpy
+```
+
+where `snforge_std::EventSpy` would allow for accessing events emitted after its creation.
+`EventSpy` would be defined as follows.
+
+```cario
+struct EventSpy {
+    events: Array<snforge_std::Event>,
+}
+
+struct Event {
+    from: ContractAddress,
+    name: felt252,
+    keys: Array<felt252>,
+    data: Array<felt252>
+}
+
+trait EventFetcher {
+    fn fetch_events(ref self: EventSpy);
+}
+
+trait EventAssertions {
+    fn assert_emitted(ref self: EventSpy, events: Array<snforge_std::Event>);
+}
+```
+
+Users will be responsible for calling `fetch_events` to load emitted events to the `events` property.
```

Comment:
Isn't it better to call it under the hood automatically every time `assert_emitted` is called? I don't see a case when someone would like to call `assert_emitted` without calling `fetch_events` first

Reply:
What if someone calls fetch_events, checks the length of the array and then calls `assert_emitted` to check if some event was emitted? We would call `fetch_events` twice